compiler/src/dotty/tools/dotc/cc/CheckCaptures.scala

```diff
@@ -729,7 +729,7 @@ class CheckCaptures extends Recheck, SymTransformer:
        *  reconstructed with the returned function.
        */
       def destructCapturingType(tp: Type, reconstruct: Type => Type = x => x): ((Type, CaptureSet, Boolean), Type => Type) =
-        tp.dealias match
+        tp.dealiasKeepAnnots match
```

Comment:
@Linyxus anything bad if I do this? `dealias` was stripping erased parameter annotations (and any other annotations)

Reply:
`tp.dealias` is causing problems if it strips all annotations. We have to change this. I'm not sure would `tp.dealiasKeepAnnots` work. It depends on the structure of annotated types, e.g. would an erased capturing argument type looks like (S @retains(cs))@erased, or the other way around?

Reply:
Currently `(S @retains(cs))@erased` is the order. I guess it depends on when the `@retains` annotations are created, vs `@erased`, but I think we can prevent the other case from happening if we surface the `@erased` annotation (if it exists) when we add the `@retains` annotation.

Reply:
If that's the case then it may not work. After `tp.dealiasKeepAnnots` it will be a type with erased annotation, which won't match the capturing type extractor. We could use `tp.dealiasKeepAnnots` and add one more case for AnnotatedType after CapturingType, and restore the annotations in `reconstruct` I think.

Reply:
Ok!